lib/ansible/modules/source_control/git.py

```diff
@@ -1019,6 +1019,14 @@ def main():
         dest = os.path.abspath(dest)
         if bare:
             gitconfig = os.path.join(dest, 'config')
+        elif os.path.isfile(os.path.join(dest, '.git')):
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+            match = re.match('gitdir: (.+)$', data)
```

Comment:
should the line start with `gitdir` or should it be in the end of line with smth prefixed? you might want to prefix the regex with `^`.

On the other hand, taking into account that regexps are expensive, you could just use str.split or partition with bits of validation on top, like:
```python
try:
    ref_prefix, separate_git_dir = data.split('gitdir: ', 1)
    if ref_prefix:
        raise ValueError('.git file has invalid git dir reference format')
    if not os.path.isdir(separate_git_dir):
        raise TypeError('%s is not a directory' % separate_git_dir)
    os.environ['GIT_DIR'] = separate_git_dir
except ValueError:
    """``.git`` file does not have a valid format for detached Git dir.""
```

Reply:
Do I have to write test cases for validations then?

Reply:
Also, what is the better way to handle the exceptions: to keep trying the execution or to shut down immediately?

Reply:
Fail fast and loud for fatal errors like being unable to fulfill the module's purpose. I've updated a snippet in the comment above to reflect this.
And yes, negative test cases for error code paths are highly appreciated, please add them.

Reply:
@cmrd-senya I've updated your branch with my implementation for now.